#536 - tests fails on windows

miroslavpojer opened the pull request:
- fixed 12 failing tests related to windows-like paths comparison

Closes #536

miroslavpojer committed:
#536 - tests fails on windows
- fixed 12 failing tests related to windows-like paths comparison

wajda commented:
Now tests are failing on Linux
```
org.scalatest.exceptions.TestFailedException: URI file:/tmp/SparkFixture3286681055411943221UnitTest/sourcetable was not the same as expected file://tmp/SparkFixture3286681055411943221UnitTest/sourcetable
```

miroslavpojer committed:
#536 - tests fails on windows
- fixed `capture lineage of 'CREATE TABLE AS` - Hive` test

miroslavpojer committed:
#536 - tests fails on windows
- fixed `capture execution error` test

miroslavpojer commented:
> Now tests are failing on Linux
> 
> ```
> org.scalatest.exceptions.TestFailedException: URI file:/tmp/SparkFixture3286681055411943221UnitTest/sourcetable was not the same as expected file://tmp/SparkFixture3286681055411943221UnitTest/sourcetable
> ```

Solved in following commits. Build server has been used to run & show these problems as I am running on windows machine.

wajda commented on integration-tests/src/test/scala/za/co/absa/spline/ExcelSpec.scala:
```diff
@@ -25,12 +25,15 @@ import za.co.absa.commons.io.{TempDirectory, TempFile}
   with SplineFixture {
 
   private val filePath = TempFile("file1", ".xlsx", false).deleteOnExit().path.toAbsolutePath.toString
+  private val fileUri = new File(filePath).toURI.toString
```
minor:
unnecessary `toString` call. The name `fileUri` suggests it is of the `URI` type, while in fact is `String`. If would be OK in other situations where it's passed to the expressions where a string is expected. But in this particular context (and the other similar ones in this PR) the given variable is only used in string interpolation, where `Any` is expected and `toString` is called implicitly. So it would be more natural for `fileUri` variable to be of the `URI` type.

cerveada replied:
maybe `toURI` method should be added to `TempFile` and `TempDirectory` so you can call 
```scala
private val fileUri = TempFile("file1", ".xlsx", false).deleteOnExit().toURI
```

wajda commented on integration-tests/src/test/scala/za/co/absa/spline/ExcelSpec.scala:
```diff
@@ -63,7 +66,7 @@ class ExcelSpec extends AsyncFlatSpec
           plan1.operations.write.append shouldBe false
           plan1.operations.write.extra.get("destinationType") shouldBe Some("excel")
-          plan1.operations.write.outputSource shouldBe s"file:$filePath"
+          plan1.operations.write.outputSource shouldBe s"$fileUri"
```
Redundant string interpolation:
```suggestion
          plan1.operations.write.outputSource shouldBe fileUri
```

miroslavpojer replied:
Agreed. Solved in following series of commit.

wajda reviewed (commented):
LGTM excepts for redundant `toString` and `s"$xxx"` as commented below.

miroslavpojer committed:
#536 - tests fails on windows
- fixed problems with Windows paths using latest chages in TempFile and TempDirectory classes
- fixed problems with embedded Derby in spark with Hive support

miroslavpojer committed:
#536 - tests fails on windows
- update of commons library version

miroslavpojer committed:
#536 - tests fails on windows
- change from asURI to toURI method name

miroslavpojer committed:
Merge branch 'develop' into bugfix/536-tests-fail-on-windows

miroslavpojer committed:
#536 - tests fails on windows
- fixed expected values create from Path instead of File

cerveada commented on integration-tests/src/test/scala/za/co/absa/spline/SQLCommandsSpec.scala:
```diff
@@ -58,9 +60,9 @@ class SQLCommandsSpec extends AsyncFlatSpec
+            plan1.operations.write.outputSource should equalToUri(s"${warehouseUri.toString}/sourcetable")
             plan2.operations.reads.get.head.inputSources.head should equalToUri(plan1.operations.write.outputSource)
-            plan2.operations.write.outputSource should equalToUri(s"file:$warehouseDir/targettable")
+            plan2.operations.write.outputSource should equalToUri(s"${warehouseUri.toString}/targettable")
```
You don't need `toString` here it's called automatically when using `s"$warehouseUri"`

miroslavpojer replied:
Fixed in commit e54f5ef9a17cf0c9ad82fbc7081d401b31a933b3.

cerveada commented on integration-tests/src/test/scala/za/co/absa/spline/harvester/LineageHarvesterSpec.scala:
```diff
@@ -52,7 +53,7 @@ class LineageHarvesterSpec extends AsyncFlatSpec
         import spark.implicits._
 
-        val testDest = tmpDest
+        val testDest = tmpDest.asString
```
In this case the code would be clearer using `tmpDest()` to make it known that method is called. I thought it was variable for some time.

Also it is recomended in scala style guide
https://docs.scala-lang.org/style/method-invocation.html#arity-0

cerveada replied:
or you could just delete the method and use the code directly it's really short and it would make it obvious what is happening.

miroslavpojer replied:
Fixed in commit 5d0f9ae49f2b674bea77d05b7807d51df7eacc06.

miroslavpojer committed:
#536 - tests fails on windows
- removed explicit to string conversion
- added usage of hive property for older Spark versions

miroslavpojer committed:
#536 - tests fails on windows
- refactoring of TempDirectory usage
- fix of failures on Spark 2.3 "derby db already used"

miroslavpojer committed:
#536 - tests fails on windows
- remove not needed property set

cerveada commented on integration-tests/src/main/scala/za/co/absa/spline/test/fixture/SparkFixture.scala:
```diff
@@ -22,21 +22,27 @@ import org.scalatest.{Assertion, AsyncTestSuite, BeforeAndAfterEach}
       .config("spark.driver.host","localhost")
       .config("spark.sql.warehouse.dir", warehouseDir)
       .config("spark.ui.enabled", "false")
+      .config("javax.jdo.option.ConnectionURL", metastoreConnectURL)
```
What is this config needed for?

miroslavpojer replied:
It is used to define custom paths to metastore_db. It solves an issue found on windows where default location was still hold by process when trying to apply clean logic.

wajda reviewed (approved):
lgtm
